### ISSUE TITLE
Probably false positive for `prefer-usememo` when `setState` is used elsewhere

### ISSUE THREAD

**Issue by Arttse** [cartant/eslint-plugin-react-etc#2]
For example:
```tsx
import { FC, useState, useCallback, useEffect } from 'react';
const SomeComponent: FC<{ data: string[] }> = ({ data }) => {
  const [state, setState] = useState<string[]>([]);

  const handleClick = useCallback(() => {
    // ...
    // there may be some other logic here
    // ...
    setState([...data, 'in handleClick']);
  }, [data]);

  // ERROR HERE:
  // ESLint: `useEffect` is forbidden here; `useMemo` should suffice.(react-etc/prefer-usememo)
  useEffect(() => {
    setState([...data, 'in useEffect']);
  }, [data]);

  return (
    <>
      <div>{state.join(', ')}</div>
      <button type="button" onClick={handleClick}>
        click me
      </button>
    </>
  );
};
```

**Comment by cartant** [comment]
I'm not sure about this. It looks like a situation that I'd like the rule to point out so that I could consider alternatives. I'll give it some thought. ATM, I'm inclined to say that what it's reporting is fine and the rule should be overridden with a comment flag where appropriate. It is, after all, a 'prefer' rule.

**Comment by Arttse** [comment]
It was a simplified example 😅 There may be more `setState` usages.

If I understand this rule correctly, then it should help when `setState` is used once. If you always force to use `useMemo`, then it seems to me too "tough" and the help from the rule is lost.

I don’t think I would use eslint disable in every such case, because my expected behavior is different and in general there can be a lot of such ignores. But maybe I'm just one of those developers who misunderstood this rule 😄 

Nevertheless, in order not to break compatibility in this case, is it possible to add an option to the rule that would allow it to pass the check when `setState` is used somewhere else (not once)?

Thank you for your time 🙌 Have a nice day

**Comment by cartant** [comment]
No worries.

The thing is, though, that you probably should use `useMemo` here even though you are not interested in any memoized value. The primary reason for the rule is that a `useEffect` call that does nothing except set a single piece of state is inefficient, as the `set` call happens after the DOM has been updated. If you were to use `useMemo` instead the `set` would be occurring _before_ the updating of the DOM. Doing that could avoid the DOM having to be updated twice.

Anyway, that's the reason for the rule and that's the reason why I think it still ought to be a warning here, too. It's possible that the rule's message doesn't convey this. I'll have a look at it later to see whether or not it can be improved.

**Comment by cartant** [comment]
The message is now a little more descriptive:

> Side effects that perform only a set-state call are inefficient; `useMemo` would be a better choice.
